core/src/main/java/com/graphhopper/reader/osm/OSMReader.java

```diff
@@ -194,23 +193,24 @@ protected boolean isBarrierNode(ReaderNode node) {
 
     /**
      * This method is called during the second pass of {@link WaySegmentParser} and provides an entry point to enrich
-     * the given OSM way with additional tags before it is split into segments.
+     * the given OSM way with additional tags before it is passed on to the tag parsers.
      */
-    protected void setArtificialWayTags(GHPoint first, GHPoint last, ReaderWay way) {
-        // todo: should we not rather do all this stuff **per edge** not per way? might be a bit slower, but we
-        // could also do raster optimization in country lookup if this was the bottleneck.
-
+    protected void setArtificialWayTags(PointList pointList, ReaderWay way) {
         // Estimate length of ways containing a route tag e.g. for ferry speed calculation
-        double firstLat = first.getLat(), firstLon = first.getLon();
-        double lastLat = last.getLat(), lastLon = last.getLon();
+        double firstLat = pointList.getLat(0), firstLon = pointList.getLon(0);
+        double lastLat = pointList.getLat(pointList.size() - 1), lastLon = pointList.getLon(pointList.size() - 1);
         GHPoint estimatedCenter = null;
+        // todonow: we have to remove the tags again, because there can be multiple edges per way, but what if this was
+        //          not an artificial tag?
+        way.removeTag("estimated_distance");
```

Comment:
maybe we start using a `gh:` prefix?

Reply:
Yes, this is a bit ugly. The cleanest thing we can do is probably using a fresh copy of the `ReaderWay` for each edge? And yes prefixing the tags to distinguish them from standard OSM tags also seems useful.

Reply:
Why do we even "abuse" tags for this? Wouldn't it be better to have a clear distinction between OSM tags and our own metadata? e.g `way.setMetadata()`

Reply:
Good point :D

Reply:
But modifying the `way` seems wrong even when we use a prefix or `way.setMetadata()`, because the artificial tags are only valid for the current edge and the way object will be used for multiple edges. Should we store the artificial 'tags' in a separate map and pass this to `handleWayTags`? So for the `EncodingManager`, `TagParser`s and `FlagEncoder` we would have something like:

```java
public IntsRef handleWayTags(ReaderWay way, Map<String, Object> edgeTags, AcceptWay acceptWay, IntsRef relationFlags)
```

?! Or should we use the `gh:` prefix **and** copy the way object for each edge?

Reply:
I would also prefer the solution with edgeTags.

Reply:
+1 for edgeTags

Reply:
It is just a bit annoying change the signatures and pass the edgeTags through to all the `TagParsers` and `FlagEncoders`. I wonder if there should be a wrapper class that holds the way, edgeTags and later nodeTags as well?

Reply:
I consider functions with more than 4 arguments to be code smell, to be honest. But I'm not sure if we should move all of them into a single wrapper class though. We could still make a distinction between "raw" data(ReaderWay) and processed data. Without having touched relation handling in GH: are the relationFlags associated with a single way or is this some shared datastructure?

Reply:
Relation flags just represent the condensed information belonging to all the relations associated to a way. The only reason we use the IntsRef/flags for this information is saving memory during import. You can think of them as data that was already extracted from the source OSM data. If we used a wrapper class, why would you prefer separating the ReaderWay from the others?

Reply:
Not really a strong preference. I was just pointing out that we *could* do this if we want to clearly separate between raw OSM data and the results of GH parsing/processing.

Reply:
I also wonder if we could postpone the signature change (even if it was the right thing to do). We will have to change the signature for #2465 and also the node tag handling might change again. Currently the only artificial/edge tag we use that actually does occur in OSM data (according to taginfo) is `country` and we already ignore/overwrite this in current master.

Reply:
Ok, let's come back to this later.